File: java/src/test/java/io/github/muthuishere/toolnexus/InProcessTest.java
package io.github.muthuishere.toolnexus;

import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicReference;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 * InProcess.createClient — a model in this JVM, with no wire configuration.
 * openspec/changes/add-in-process-client. Mirrored in all seven ports.
 */
class InProcessTest {

    private static Tool addTool() {
        return NativeTool.of("add", "Add two numbers.",
                Map.of("type", "object",
                        "properties", Map.of("a", Map.of("type", "number"), "b", Map.of("type", "number")),
                        "required", List.of("a", "b")),
                args -> String.valueOf((long) (((Number) args.get("a")).doubleValue()
                        + ((Number) args.get("b")).doubleValue())));
    }

    @Test
    void noWireConfigurationIsRequired() throws IOException {
        Toolkit tk = Toolkit.create(new Toolkit.Options());
        // No baseUrl. No apiKey. No style. That is the whole point.
        LlmClient client = InProcess.createClient(new InProcess.Options()
                .model("my-local")
                .generate(req -> InProcess.Response.content("hello from in-process")));

        LlmClient.RunResult r = client.run("hi", tk);
        assertEquals("hello from in-process", r.text);
        assertEquals("done", r.status);
        tk.close();
    }

    @Test
    void generateSeesTheAssembledRequest() throws IOException {
        Toolkit tk = Toolkit.create(new Toolkit.Options());
        tk.register(addTool());
        AtomicReference<InProcess.Request> seen = new AtomicReference<>();

        LlmClient client = InProcess.createClient(new InProcess.Options()
                .model("my-local").systemPrompt("You are terse.")
                .generate(req -> { seen.set(req); return InProcess.Response.content("ok"); }));
        client.run("What is 2 + 3?", tk);

        assertEquals("my-local", seen.get().model);
        assertTrue(seen.get().tools.size() >= 1, "tool schemas are offered");
        String messages = Json.stringify(seen.get().messages);
        assertTrue(messages.contains("terse"), messages);
        assertTrue(messages.contains("2 + 3"), messages);
        tk.close();
    }

    @Test
    void toolCallsLoopBackWithTheResult() throws IOException {
        Toolkit tk = Toolkit.create(new Toolkit.Options());
        tk.register(addTool());
        AtomicInteger n = new AtomicInteger();

        LlmClient client = InProcess.createClient(new InProcess.Options().model("m")
                .generate(req -> n.incrementAndGet() == 1
                        ? InProcess.Response.toolCalls(new InProcess.ToolCall("add", Map.of("a", 2, "b", 3)))
                        : InProcess.Response.content("the answer is 5")));

        LlmClient.RunResult r = client.run("What is 2 + 3?", tk);
        assertEquals(1, r.toolCalls.size());
        assertEquals("add", r.toolCalls.get(0).name);
        assertEquals("5", r.toolCalls.get(0).output);
        tk.close();
    }

    @Test
    void argumentsStructuredOrPreEncoded() throws IOException {
        for (Object args : List.of(Map.of("a", 2, "b", 3), "{\"a\":2,\"b\":3}")) {
            Toolkit tk = Toolkit.create(new Toolkit.Options());
            tk.register(addTool());
            AtomicInteger n = new AtomicInteger();
            LlmClient client = InProcess.createClient(new InProcess.Options().model("m")
                    .generate(req -> n.incrementAndGet() == 1
                            ? InProcess.Response.toolCalls(new InProcess.ToolCall("add", args))
                            : InProcess.Response.content("done")));
            LlmClient.RunResult r = client.run("go", tk);
            assertEquals("5", r.toolCalls.get(0).output, "args form: " + args);
            tk.close();
        }
    }

    @Test
    void usageIsOptionalAndDerived() throws IOException {
        Toolkit tk = Toolkit.create(new Toolkit.Options());

        LlmClient bare = InProcess.createClient(new InProcess.Options().model("m")
                .generate(req -> InProcess.Response.content("x")));
        assertEquals(0, bare.run("hi", tk).usage.totalTokens, "absent usage is zero, not a failure");

        LlmClient counted = InProcess.createClient(new InProcess.Options().model("m")
                .generate(req -> InProcess.Response.content("x").usage(11, 4)));
        LlmClient.RunResult r = counted.run("hi", tk);
        assertEquals(11, r.usage.promptTokens);
        assertEquals(15, r.usage.totalTokens, "total is derived when not given");
        tk.close();
    }

    @Test
    void streamingIsRefusedLoudly() throws IOException {
        Toolkit tk = Toolkit.create(new Toolkit.Options());
        LlmClient client = InProcess.createClient(new InProcess.Options().model("m")
                .generate(req -> InProcess.Response.content("x")));

        StringBuilder text = new StringBuilder();
        Exception boom = assertThrows(Exception.class,
                () -> client.stream("hi", tk, ev -> {
                    if (ev.type() == LlmClient.StreamEvent.Kind.TEXT) text.append(ev.delta());
                }));
        assertTrue(String.valueOf(boom.getMessage()).contains("does not support streaming")
                        || String.valueOf(boom.getCause()).contains("does not support streaming"),
                "message was: " + boom);
        assertEquals(0, text.length(), "no delta may be emitted before refusing");
        tk.close();
    }

    @Test
    void failuresAreNotRetriedByDefault() throws IOException {
        // There is no wire, so there is no transient failure to ride out: a generate that
        // fails will fail again, and retrying only buys backoff over the caller's own bug.
        Toolkit tk = Toolkit.create(new Toolkit.Options());
        AtomicInteger calls = new AtomicInteger();
        LlmClient client = InProcess.createClient(new InProcess.Options().model("m")
                .generate(req -> { calls.incrementAndGet(); throw new IllegalStateException("my model blew up"); }));

        long started = System.nanoTime();
        assertThrows(Exception.class, () -> client.run("hi", tk));
        assertEquals(1, calls.get(), "a local failure is not transient");
        assertTrue((System.nanoTime() - started) / 1_000_000 < 500, "must fail immediately");
        tk.close();
    }

    @Test
    void generateIsRequired() {
        assertThrows(IllegalArgumentException.class,
                () -> InProcess.createClient(new InProcess.Options().model("m")));
    }
}

File: golang/inprocess_test.go
package toolnexus

import (
	"context"
	"encoding/json"
	"strconv"
	"strings"
	"testing"
)

// CreateInProcessClient — a model in this process, with no wire configuration.
// openspec/changes/add-in-process-client. Mirrored in all seven ports.

func addTool(t *testing.T) Tool {
	t.Helper()
	return NativeTool("add", "Add two numbers.",
		JSONSchema{"type": "object", "properties": map[string]any{
			"a": map[string]any{"type": "number"}, "b": map[string]any{"type": "number"}},
			"required": []string{"a", "b"}},
		func(_ context.Context, args map[string]any) (string, error) {
			a, _ := args["a"].(float64)
			b, _ := args["b"].(float64)
			return strconv.FormatFloat(a+b, 'f', -1, 64), nil
		})
}

func bareToolkit(t *testing.T) *Toolkit {
	t.Helper()
	tk, err := CreateToolkit(context.Background(), Options{Builtins: false})
	if err != nil {
		t.Fatal(err)
	}
	return tk
}

func TestInProcessNoWireConfiguration(t *testing.T) {
	tk := bareToolkit(t)
	defer tk.Close()
	// No BaseURL. No APIKey. No Style. That is the whole point.
	c := CreateInProcessClient(InProcessOptions{
		Model: "my-local",
		Generate: func(InProcessRequest) (InProcessResponse, error) {
			return InProcessResponse{Content: "hello from in-process"}, nil
		},
	})
	r, err := c.Run(context.Background(), "hi", tk)
	if err != nil || r.Text != "hello from in-process" || r.Status != "done" {
		t.Fatalf("got %+v err=%v", r, err)
	}
}

func TestInProcessGenerateSeesTheAssembledRequest(t *testing.T) {
	tk, err := CreateToolkit(context.Background(), Options{Builtins: false})
	if err != nil {
		t.Fatal(err)
	}
	defer tk.Close()
	tk.Register(addTool(t))

	var seen InProcessRequest
	c := CreateInProcessClient(InProcessOptions{
		Model: "my-local", SystemPrompt: "You are terse.",
		Generate: func(req InProcessRequest) (InProcessResponse, error) {
			seen = req
			return InProcessResponse{Content: "ok"}, nil
		},
	})
	if _, err := c.Run(context.Background(), "What is 2 + 3?", tk); err != nil {
		t.Fatal(err)
	}
	if seen.Model != "my-local" {
		t.Fatalf("model = %q", seen.Model)
	}
	if len(seen.Tools) == 0 {
		t.Fatal("tool schemas were not offered")
	}
	blob, _ := json.Marshal(seen.Messages)
	if !strings.Contains(string(blob), "terse") || !strings.Contains(string(blob), "2 + 3") {
		t.Fatalf("messages missing the system prompt or the user turn: %s", blob)
	}
}

func TestInProcessToolCallsLoopBack(t *testing.T) {
	tk, _ := CreateToolkit(context.Background(), Options{Builtins: false})
	defer tk.Close()
	tk.Register(addTool(t))

	n := 0
	c := CreateInProcessClient(InProcessOptions{Model: "m",
		Generate: func(req InProcessRequest) (InProcessResponse, error) {
			n++
			if n == 1 {
				return InProcessResponse{ToolCalls: []InProcessToolCall{
					{Name: "add", Arguments: map[string]any{"a": 2, "b": 3}}}}, nil
			}
			return InProcessResponse{Content: "the answer is 5"}, nil
		}})
	r, err := c.Run(context.Background(), "What is 2 + 3?", tk)
	if err != nil {
		t.Fatal(err)
	}
	if len(r.ToolCalls) != 1 || r.ToolCalls[0].Name != "add" || r.ToolCalls[0].Output != "5" {
		t.Fatalf("tool calls = %+v", r.ToolCalls)
	}
}

func TestInProcessArgumentsStructuredOrPreEncoded(t *testing.T) {
	for _, args := range []any{map[string]any{"a": 2, "b": 3}, `{"a":2,"b":3}`} {
		tk, _ := CreateToolkit(context.Background(), Options{Builtins: false})
		tk.Register(addTool(t))
		n := 0
		c := CreateInProcessClient(InProcessOptions{Model: "m",
			Generate: func(InProcessRequest) (InProcessResponse, error) {
				n++
				if n == 1 {
					return InProcessResponse{ToolCalls: []InProcessToolCall{{Name: "add", Arguments: args}}}, nil
				}
				return InProcessResponse{Content: "done"}, nil
			}})
		r, err := c.Run(context.Background(), "go", tk)
		if err != nil || len(r.ToolCalls) != 1 || r.ToolCalls[0].Output != "5" {
			t.Fatalf("args %v -> %+v err=%v", args, r.ToolCalls, err)
		}
		tk.Close()
	}
}

func TestInProcessUsageIsOptional(t *testing.T) {
	tk := bareToolkit(t)
	defer tk.Close()

	bare := CreateInProcessClient(InProcessOptions{Model: "m",
		Generate: func(InProcessRequest) (InProcessResponse, error) {
			return InProcessResponse{Content: "x"}, nil
		}})
	r, _ := bare.Run(context.Background(), "hi", tk)
	if r.Usage.TotalTokens != 0 {
		t.Fatalf("absent usage should be zero, got %d", r.Usage.TotalTokens)
	}

	counted := CreateInProcessClient(InProcessOptions{Model: "m",
		Generate: func(InProcessRequest) (InProcessResponse, error) {
			return InProcessResponse{Content: "x",
				Usage: &InProcessUsage{PromptTokens: 11, CompletionTokens: 4}}, nil
		}})
	r2, _ := counted.Run(context.Background(), "hi", tk)
	if r2.Usage.PromptTokens != 11 || r2.Usage.TotalTokens != 15 {
		t.Fatalf("usage = %+v; total should be derived", r2.Usage)
	}
}

func TestInProcessStreamingIsRefusedLoudly(t *testing.T) {
	tk := bareToolkit(t)
	defer tk.Close()
	c := CreateInProcessClient(InProcessOptions{Model: "m",
		Generate: func(InProcessRequest) (InProcessResponse, error) {
			return InProcessResponse{Content: "x"}, nil
		}})
	ch, err := c.Stream(context.Background(), "hi", tk)
	if err != nil {
		if !strings.Contains(err.Error(), "does not support streaming") {
			t.Fatalf("wrong error: %v", err)
		}
		return
	}
	// If the call itself did not fail, the error must arrive on the channel — and no
	// content delta may precede it, or we faked a stream.
	for ev := range ch {
		if ev.Type == "text" {
			t.Fatal("emitted a text delta instead of refusing")
		}
		if ev.Err != nil && strings.Contains(ev.Err.Error(), "does not support streaming") {
			return
		}
	}
	t.Fatal("streaming was neither refused nor errored")
}
